### PR TITLE
refactor: CLI client interface

### LINKED ISSUE #1472: Symmetry of RPC and HTTP APIs
DefraDB offers HTTP and RPC APIs to do the non-GraphQL DB operations. The HTTP one is necessary to provide a conventional GraphQL endpoint.

There should be symmetry between RPC and HTTP APIs. It would allow more clarity and flexibility of API consumption. Additionally, DefraDB's language clients would be simplified to mostly a thin layer over autogenerated gRPC code.

Note 1: It may be additionally possible to offer the GraphQL endpoint via RPC as well. This can be another issue after this one.
Note 2: Offering documentation for the RPC API on the documentation website can be tracked in another issue. The issue to track HTTP documentation is https://github.com/sourcenetwork/defradb/issues/510
Note 3: This excludes p2p RPC surface area.

### LINKED ISSUE #1507: Complete RPC API
Complete the RPC API by introducing following methods (or something similar / more well-typed):

```
    rpc Ping(PingRequest) returns (PingReply) {}
    rpc PeerID (PeerIDRequest) returns (PeerIDReply) {}
    rpc SchemaLoad(SchemaLoadRequest) returns (SchemaLoadReply) {}
    rpc SchemaPatch(SchemaPatchRequest) returns (SchemaPatchReply) {}
    rpc GraphQL(GraphQLRequest) returns (GraphQLReply) {}
```

Part of https://github.com/sourcenetwork/defradb/issues/1472
Ideally would be after / depend-on https://github.com/sourcenetwork/defradb/issues/1063

### LINKED ISSUE #1860: Rework code coverage action(s) to allow all test jobs to contribute

### PR DESCRIPTION
**Pull request by nasdf**
## Relevant issue(s)

Closes #1472
Closes #1507
Closes #1860 

## Description

This is a follow up to #1776

This PR adds a CLI implementation that implements the client.DB interface and runs through the existing integration test suite.

- [x] Merge existing server config code
- [x] Refactor CLI to use new HTTP client
- [x] Remove `net/api` package
- [x] Remove `api/http` package
- [x] Lens tests are timing out in CI: fixed #1862
- [x] Code coverage is incorrectly reporting: fixed #1861
- [x] Flaky test causing failures: fixed #1912

Renamed Commands:
- `peerid` to `client peer info`
- `client p2pcollection` to `client p2p collection`
- `client replicator` to `client p2p replicator`
- `client schema list` to `client collection describe`

Removed Commands:
- `block get`
- `ping`
- `rpc`

Added Commands:
- `client collection create`
- `client collection delete`
- `client collection get`
- `client collection keys`
- `client collection update`
- `client tx create`
- `client tx discard`
- `client tx commit`
- `client schema migration up`
- `client schema migration down`
- `client schema migration reload`

**Notes for reviewers**:
- `.github` changes are merged from #1871
- `Makefile` most of these changes are also from #1871
- `docs/cli` ignore these changes, it will be updated next release
- sorry for all of the merge commits, I am working on learning rebase flow

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

`make test`

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: http/handler.go
Comment (nasdf) [inline http/handler.go:1]: All of this was pulled from `server.go` to have a cleaner separation of routes and server setup logic.

File: http/server.go
Comment (nasdf) [inline http/server.go:1]: This was essentially copied over from `api/http/server` along with the original tests.

File: logging/registry.go
Comment (nasdf) [inline logging/registry.go:48]: This is another race condition that I discovered.

File: http/errors.go
Comment (nasdf) [inline http/errors.go:1]: In here I merged in errors from `api/http/errors.go`

File: http/client_tx.go
Comment (nasdf) [inline http/client_tx.go:29]: This was added so the CLI tests have access to the transactions. This should be cleaned up when we switch to an internal transaction pool.

File: http/handler.go
Comment (nasdf) [inline http/handler.go:128]: This was added so the CLI tests have access to the transactions. This should be cleaned up when we switch to an internal transaction pool.

File: cli/wrapper.go
Comment (nasdf) [inline cli/wrapper.go:1]: The wrapper translates `client.DB` functions to CLI calls. It's currently a bit more work than I'd like when adding new functions, but this is the trade off instead of writing unit tests for each implementation.

File: Makefile
Comment (nasdf) [inline Makefile:92]: `make test:ci` now includes coverage

File: Makefile
Comment (shahzadlone) [inline Makefile:193]: question: wouldn't this mean that all matrix instances will now build with lens dependencies (even if they don't need lens deps).

Edit: I noticed now we are actually always going to run all lens tests so I guess it fine, I initially thought the lens test will run separately as a row splice in a matrix. Look at suggestion below in that case.

File: Makefile
Comment (shahzadlone) [inline Makefile:192]: todo: suggest this should not be named as `make test:ci` anymore as it doesn't make sense anymore. Perhaps something like `make coverage`. You can then reuse this make rule to replace the first 5 lines of `test\:coverage:` rule as well, it's now generating code coverage of all the tests (of course and running all tests too).

File: Makefile
Comment (shahzadlone) [inline Makefile:268]: suggestion: Can be simplified perhaps if the above suggestion is applied.

```suggestion
	@$(MAKE) coverage
	go tool covdata textfmt -i=$(COVERAGE_DIRECTORY) -o $(COVERAGE_FILE)
```

File: cmd/defradb/main.go
Comment (shahzadlone) [inline cmd/defradb/main.go:22]: question: why?

**Comment by AndrewSisley** [comment]
Can I please request that next time you do not merge multiple significant changes (i.e. the test matrix stuff, and the cli stuff) into one giant PR. Giant PRs are not fun to review.  

It increases the risk of reviewers (and authors) missing stuff - the PR lacks focus and participants will experience greater levels of fatigue working through it. 

It prevents useful changes from being merged, as comments on one change will block the other change from being merged (preventing users and Source developers from using them).  This also maximises the disruption caused when you do finally merge the one PR, there are more points of conflict, and more changes other developers will be forced to learn and incorporate into their own branches in one go.

It increases the likelihood that discussions will be cut short in order to 'just get it in'.

It also results in stuff being hidden in the main/develop commit history, as instead of having commits `Foo` and `Bar`, we just have one giant `FooBar` commit.  This also makes rolling back of `Foo` quite difficult, unless `Bar` is also rolled back with it.

By merging together PRs like this you are forcing yourself to have multiple WIP items at the same time - everything is in progress, but nothing is being delivered - kind of like the old waterfall development flow.

File: cmd/defradb/main.go
Comment (nasdf) [reply cmd/defradb/main.go]: I've updated it to exit with a non-zero status code on error. The errors are already logged by cobra.

File: Makefile
Comment (nasdf) [reply Makefile]: I've merged the coverage tests like you suggested.

**Comment by nasdf** [comment]
> Can I please request that next time you do not merge multiple significant changes (i.e. the test matrix stuff, and the cli stuff) into one giant PR. Giant PRs are not fun to review.

I wouldn't normally take this approach but there were two issues that forced me to merge them:

1. The test matrix PR fails due to flaky cli tests
2. This PR will timeout in CI without the new test matrix

> It increases the risk of reviewers (and authors) missing stuff - the PR lacks focus and participants will experience greater levels of fatigue working through it.

I also left the original test matrix PR open in case anyone wants to review them separately.

> It prevents useful changes from being merged, as comments on one change will block the other change from being merged (preventing users and Source developers from using them). This also maximises the disruption caused when you do finally merge the one PR, there are more points of conflict, and more changes other developers will be forced to learn and incorporate into their own branches in one go.

That's the case for any large change.

> It increases the likelihood that discussions will be cut short in order to 'just get it in'.

This is a problem with any code review.

> It also results in stuff being hidden in the main/develop commit history, as instead of having commits `Foo` and `Bar`, we just have one giant `FooBar` commit. This also makes rolling back of `Foo` quite difficult, unless `Bar` is also rolled back with it.

The commit histories are squashed and merged so its really just one vs two commits here.

> By merging together PRs like this you are forcing yourself to have multiple WIP items at the same time - everything is in progress, but nothing is being delivered - kind of like the old waterfall development flow.

I work on multiple things all the time. I think it's the nature of an async team.

**Comment by AndrewSisley** [comment]
I've moved this conversation to discord, thread `dev-db.Large PRs` https://discord.com/channels/427944769851752448/1157020013828509746

**Comment by shahzadlone** [comment]
@nasdf Reviewing the changes and the remaining PR, will give LGTM once done.

Also I don't think I have god powers any more so @jsimnz might need to remove the 3 required workflows that morphed into the matrix workflow, inorder for this to merge.

File: .github/workflows/run-tests-and-upload-coverage.yml
Comment (shahzadlone) [inline .github/workflows/run-tests-and-upload-coverage.yml:19]: todo: I think this workflow is expensive to run on every push (x2 for every push on branch flow PRs). 

```suggestion
  push:
    tags:
      - 'v[0-9]+.[0-9]+.[0-9]+'
    branches:
      - master
      - develop
```

File: .github/workflows/run-tests-and-upload-coverage.yml
Comment (shahzadlone) [inline .github/workflows/run-tests-and-upload-coverage.yml:11]: todo: The name isn't updated.

File: .github/workflows/run-tests-and-upload-coverage.yml
Comment (AndrewSisley) [reply .github/workflows/run-tests-and-upload-coverage.yml:19]: I disagree with you here Shahzad - I want these tests to run, and pass, before a PR can be merged.

File: cmd/defradb/main.go
Comment (AndrewSisley) [reply cmd/defradb/main.go]: Can you please add a code-comment documenting why you think this error is fine to discard.

File: cmd/defradb/main.go
Comment (AndrewSisley) [inline cmd/defradb/main.go:24]: todo: Similar to other comment, please add a code comment documenting why it is fine to discard this error

File: cli/wrapper.go
Comment (AndrewSisley) [inline cli/wrapper.go:33]: question: Why is this production code?

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
I'm only part way through my review, but I thought I'd submit the comments I have so far. Overall it looks really good, and it is a fantastic amount of code you've been able to remove, whilst gaining features :)

File: cli/document_keys.go
Comment (AndrewSisley) [inline cli/document_keys.go:52]: todo: Please document why this is okay

File: cli/p2p_replicator_delete.go
Comment (AndrewSisley) [inline cli/p2p_replicator_delete.go:26]: todo: This entire command appears to be untested, please test it.

File: cli/p2p_replicator_set.go
Comment (AndrewSisley) [inline cli/p2p_replicator_set.go:23]: todo: `[-c, --collection]` appears to be inconsistent with the rest of the CLI, I think it should be consistent.

File: cli/peerid.go
Comment (AndrewSisley) [inline cli/peerid.go:26]: question: Is this now provided by p2p_info?

File: cli/schema_migration_down.go
Comment (AndrewSisley) [inline cli/schema_migration_down.go:27]: todo: Please explain to the user what `<src>` is :)

suggestion: `on` is somewhat ambiguous here I think, `from` (or `to`, I cant remember which way round) might be better.
 
 Same goes for the up command.

**Review by AndrewSisley — CHANGES_REQUESTED** [review]

File: cli/blocks_get.go
Comment (nasdf) [reply cli/blocks_get.go:24]: I've added a list of the new, renamed, and removed commands to the description.

File: .github/workflows/run-tests-and-upload-coverage.yml
Comment (shahzadlone) [reply .github/workflows/run-tests-and-upload-coverage.yml:19]: > I disagree with you here Shahzad - I want these tests to run, and pass, before a PR can be merged. I also want accurate codecov within a PR.

Where is the disagreement? Not only should these run, and pass in a PR, this should also be a required / mandatory check. Nowhere in my suggestion I suggested otherwise.

Note - I am not saying to remove this part:
```
  pull_request:
    branches:
      - master
      - develop
```

All I am saying is we shouldn't have this run on every branch, in addition to the PR, and avoid running twice on PRs opened by branch flow (i.e. dependabot PRs will run once for the PR and once for the push).

File: cli/wrapper.go
Comment (nasdf) [reply cli/wrapper.go:33]: I'd like to move this and the http wrapper to the integration tests package. I think there was something blocking me from doing that, but I will give it another look.

File: cli/peerid.go
Comment (nasdf) [reply cli/peerid.go:26]: Yes, as a follow up I will be adding the multiaddresses as a return value there as well.

File: .github/workflows/run-tests-and-upload-coverage.yml
Comment (AndrewSisley) [reply .github/workflows/run-tests-and-upload-coverage.yml:19]: > Not only should these run, and pass in a PR, this should also be a required / mandatory check. 

Ah! Sorry, I thought your snippet was saying that they should only run on those branches/tags, and not in PR :) Thanks for the clarification.

File: cli/blocks_get.go
Comment (AndrewSisley) [reply cli/blocks_get.go:24]: Cheers Keenan!

File: cli/wrapper.go
Comment (AndrewSisley) [reply cli/wrapper.go:33]: Cheers Keenan - please give me a shout here once you've pushed changes/decided against moving and I'll review them fully.

File: cli/document_create.go
Comment (nasdf) [reply cli/document_create.go:33]: I've reorganized this so it better matches the client API.

File: cli/document_create.go
Comment (nasdf) [reply cli/document_create.go:76]: I think adding more integration tests would be the best option. I will put together a list of areas that are missing coverage.

File: cli/document_keys.go
Comment (nasdf) [reply cli/document_keys.go:52]: I decided to return the error here instead.

File: cli/p2p_replicator_delete.go
Comment (nasdf) [reply cli/p2p_replicator_delete.go:26]: I'd like for this to be covered in an integration test so we can get coverage on other implementations as well. I'll follow up shortly with a fix.

File: cli/p2p_replicator_set.go
Comment (nasdf) [reply cli/p2p_replicator_set.go:23]: Are you referring to the naming or that it's a slice type?

File: cli/wrapper.go
Comment (nasdf) [reply cli/wrapper.go:33]: Update: I was able to move them to a new `tests/clients` package.

File: cli/schema_migration_down.go
Comment (nasdf) [reply cli/schema_migration_down.go]: I tried to describe it in more detail. I believe it is the list of documents to apply / reverse the migration to / from.

File: cli/collection.go
Comment (AndrewSisley) [inline cli/collection.go:32]: Sweet :) I was hoping you'd be able to find a way to do it like this when I was commenting - forces consistency and we get all the features :)

File: cli/collection_create.go
Comment (AndrewSisley) [inline cli/collection_create.go:37]: suggestion: Similar to `store`, I'd suggest refactoring out `cmd.Context().Value(colContextKey).(client.Collection)` into a private func, located next to where it is set (collection.go).

File: cli/p2p_replicator_set.go
Comment (AndrewSisley) [reply cli/p2p_replicator_set.go:23]: I forgot that this was a slice type, but I was referring to the existence of the shorthand.  However the way you have moved the document stuff into collection has removed the inconsistency (as there is no longer a `--collection` flag there)

File: cli/schema_migration_down.go
Comment (AndrewSisley) [reply cli/schema_migration_down.go]: cheers Keenan - looks good

**Review by AndrewSisley — APPROVED** [review]
LGTM, and looking forward to having it in :)  Would suggest a quick manual sanity check of a couple of CLI commands before merge, just in case :)

Good job Keenan :)

**Review by shahzadlone — APPROVED** [review]
Cheers! Just posted one small nitpick.

We can coordinate merging this as would need to turn off the required checks right before merge.

**Pull request merged**
